Cython/Utility/ExtensionTypes.c

```diff
@@ -155,38 +155,36 @@ static int __Pyx_validate_bases_tuple(const char *type_name, Py_ssize_t dictoffs
 #endif
             return -1;
         }
-#if !CYTHON_USE_TYPE_SLOTS
-        if (dictoffset == 0) {
-            PyErr_Format(PyExc_TypeError,
-                "extension type '%s.200s': "
-                "unable to validate whether bases have a __dict__ "
-                "when CYTHON_USE_TYPE_SLOTS is off "
-                "(likely because you are building in the limited API). "
-                "Therefore, all extension types with multiple bases "
-                "must add 'cdef dict __dict__' in this compilation mode",
-                type_name);
-#if CYTHON_AVOID_BORROWED_REFS
-            Py_DECREF(b0);
-#endif
-            return -1;
-        }
-#else
-        if (dictoffset == 0 && b->tp_dictoffset)
+        if (dictoffset == 0)
         {
-            __Pyx_TypeName b_name = __Pyx_PyType_GetName(b);
-            PyErr_Format(PyExc_TypeError,
-                "extension type '%.200s' has no __dict__ slot, "
-                "but base type '" __Pyx_FMT_TYPENAME "' has: "
-                "either add 'cdef dict __dict__' to the extension type "
-                "or add '__slots__ = [...]' to the base type",
-                type_name, b_name);
-            __Pyx_DECREF_TypeName(b_name);
-#if CYTHON_AVOID_BORROWED_REFS
-            Py_DECREF(b0);
-#endif
-            return -1;
+            Py_ssize_t b_dictoffset = 0;
+#if CYTHON_USE_TYPE_SLOTS
+            b =  = b->tp_dictoffset;
+#else
+            PyObject *py_b_dictoffset = PyObject_GetAttrString((PyObject*)b, "__dictoffset__");
+            if (!py_b_dictoffset) goto dictoffset_return;
+            b_dictoffset = PyLong_AsSsize_t(py_b_dictoffset);
+            Py_DECREF(py_b_dictoffset);
+            if (b_dictoffset == -1 && PyErr_Occurred()) goto dictoffset_return;
+#endif
+            if (b_dictoffset) {
+                __Pyx_TypeName b_name = __Pyx_PyType_GetName(b);
```

Comment:
The declaration of `b_name` is between the `goto` and the label, which will fail in C++

Reply:
Adding another set of braces to enclose the scope of `b_name` fixes the c++ compilation error, but then running the test on PyPy I get
```
CFLAGS="-DCYTHON_USE_TYPE_SPECS=1" ../pypy/pypy3.9-c runtests.py cdef_multiple_inheritance_nodict --no-c --no-cleanup
...
File "tests/run/cdef_multiple_inheritance_nodict.pyx", line 16, in init cdef_multiple_inheritance_nodict (cdef_multiple_inheritance_nodict.cpp:7472)
    cdef class Both(CBase, PyBase):
AttributeError: type object 'PyBase' has no attribute '__dictoffset__'
```

which seems to be an incompatibility in PyPy:  classes never have `__dictoffset__`. It turns out there is an [open issue](https://foss.heptapod.net/pypy/pypy/-/issues/2447) about the lack of proper handling of `_PyObject_GetDictPtr` and `tp_dictoffset` for pybind11, that was never resolved. So while this PR fixes a corner case for CPython, it doesn't fix the PyPy test failure.

Sorry for the bother. I am not sure how

Reply:
well, I guess I could get `USE_CYTHON_TYPE_SLOTS` to work on PyPy.

Reply:
I made the test

```
#if CYTHON_USE_TYPE_SLOTS || CYTHON_COMPILING_IN_PYPY
```

Not completely sure if this'll work or not though, but it seems easier than trying to get the *whole* of CYTHON_USE_TYPE_SLOTS to work with PyPy.

> Sorry for the bother. I am not sure how to fix this.

It's still worth fixing the limited api side of it anyway, even if it doesn't fix PyPy.